### DISCUSSION TITLE
Block Coding with optional deferred / lazy evalution of argument for own tasks

### DISCUSSION THREAD

**Discussion by ggramlich** [pybricks/support#1325] [Ideas]
I would like to use a pattern where I pass conditions (boolean functions) to functions as lambdas without arguments. This works nicely for the Python version and I think, it might be possible to introduce it to Block Coding.

Example with a line follower that starts out slowly to pick up the line for the first 100mm and then speeds up, until it recognizes a black crossing.

```
def follow_until(speed: int, kp: float, condition: Callable[[], any]):
    while not condition():
        robot.drive(speed, kp * (target_reflection - color_sensor.reflection()))

# The main program starts here.
follow_until(200, 1.2, lambda: (robot.distance() > 100))
follow_until(400, 0.5, lambda: (color_sensor.reflection() < 30))
robot.brake()
```

For the blocks you could define it on the specific argument of the own task that is expected to be evaluated deferred.
The generated Python code would then wrap the argument on the function call with `lambda: ...` and would add `()` to the usage of the argument in the body of the function.

Example usage in Blocks
![PybricksBlocksDeferedEvaluationProposal](https://github.com/pybricks/support/assets/69276/60328e0d-244a-4805-b348-9619d95ff7f6)

Example with a task for the condition itself
![PybricksBlocksDeferedEvaluationProposal_OnBlackTask](https://github.com/pybricks/support/assets/69276/c562c651-daf9-4994-80d8-86ace9bd4273)

This is not a necessity to me and of course we would find our way around it in Blocks, but would make for a nice addition.

**Comment by laurensvalk** [comment]
Thanks for your suggestion. It does seem fairly advanced, but now that my mind is at it, I'll leave some notes here on one way how it might work, without additional options on existing blocks.

Basically, in your code - you could think of having a new block in exactly the same shape of your red outline. The code generation would work as follows:
- The blocks inside your red lines would be treated as if wrapped in a normal function definition block. So it would create:
```python
def func1():
    return condition
```
- The red outline itself would just generate `func1`. It's just an object that can be passed as an argument as in your case.
- There is already a block to call callables, so you could use the object with that.

Doing it this way doesn't require `lambda` and makes it a bit more flexible as you can pass around the callable or store it in an object.

Long term, I'm certainly not against something like this, especially if it's only an extra block that doesn't affect normal usage. I'm noticing a trend where people are quite interested in a high-ceiling block language (even coming from regular MicroPython folks), so maybe this fits well in that direction...

**Reply by ggramlich** [reply]
Thanks @laurensvalk for considering it.

Do I understand correctly that to implement it you think about a new type of "bubble" block into which you could put another bubble?

I did find a way around it, which already works. But **I do not recommend that to anybody**, nor would I show that to the kids.
It uses reflection and allows to call an "own task" without arguments by its name as string like so. You can also make something similar work for own tasks with arguments.

File: deferred.py
```py
def call_by_name(fn_name: string):
    # print('calling', fn_name)
    return locals()[fn_name]()
```

and you use it like this:
![image](https://github.com/pybricks/support/assets/69276/4e1b92b1-0203-458a-b77b-585e984c5e91)

-------------
Another way might be

File: deferred.py
```py
def evaluate(fn: callable):
    return fn()

def wrap(fn_name: string):
    return locals()[fn_name]
```

![image](https://github.com/pybricks/support/assets/69276/ce65bff9-881a-405a-bb7a-8106e8508e9e)
